src/MLJCatBoostInterface.jl

```diff
@@ -0,0 +1,159 @@
+const MMI = MLJModelInterface
+using MLJModelInterface: Table, Continuous, Count, Finite, OrderedFactor, Multiclass
+const PKG = "CatBoost"
+
+"""
+    mlj_to_kwargs(model)
+
+Convert MLJ model struct to a dict of kwargs
+
+Parameters
+----------
+- `model`: MLJ Model
+
+Returns
+-------
+- `Dict{Symbol, Any}`:Dictionary of kwargs
+"""
+function mlj_to_kwargs(model)
+    return Dict{Symbol,Any}(name => getfield(model, name)
+                            for name in fieldnames(typeof(model)))
+end
+
+"""
+    get_dtype_feature_ix(X, dtype)
+
+Get the index of the columns with a specific dtype
+
+Parameters
+----------
+- `X`: Table
+- `dtype`: DataType 
+
+Returns
+-------
+- `Vector{Int64}`
+"""
+function get_dtype_feature_ix(X, dtype)
+    return findall(MMI.schema(X).scitypes .<: dtype)
+end
+
+function drop_cols(a::NamedTuple{an}, cols::Tuple) where {an}
+    names = Base.diff_names(an, cols)
+    return NamedTuple{names}(a)
+end
+
+function prepare_input(X, y)
+    table_input = Tables.columntable(X)
+    columns = Tables.columnnames(table_input)
+
+    order_factor_cols = columns[get_dtype_feature_ix(table_input, OrderedFactor)]
+    new_columns = NamedTuple{order_factor_cols}([MMI.int(table_input[col])
+                                                 for col in order_factor_cols])
+    table_input = (; drop_cols(table_input, order_factor_cols)..., new_columns...)
+
+    cat_features = get_dtype_feature_ix(table_input, Multiclass) .- 1 # convert to 0 based indexing
+    text_features = get_dtype_feature_ix(table_input, MMI.Textual) .- 1 # convert to 0 based indexing
+    data_pool = Pool(table_input; label=numpy.array(Array(y)), cat_features, text_features)
+
+    return data_pool
+end
+
+function prepare_input(X)
+    table_input = Tables.columntable(X)
+    columns = Tables.columnnames(table_input)
+
+    order_factor_cols = columns[get_dtype_feature_ix(table_input, OrderedFactor)]
+    new_columns = NamedTuple{order_factor_cols}([MMI.int(table_input[col])
+                                                 for col in order_factor_cols])
+    table_input = (; drop_cols(table_input, order_factor_cols)..., new_columns...)
+
+    cat_features = get_dtype_feature_ix(table_input, Multiclass) .- 1 # convert to 0 based indexing
+    text_features = get_dtype_feature_ix(table_input, MMI.Textual) .- 1 # convert to 0 based indexing
+
+    X_pool = Pool(table_input; cat_features, text_features)
+
+    return X_pool
+end
+
+include("mlj_catboostclassifier.jl")
+include("mlj_catboostregressor.jl")
+
+const CatBoostModels = Union{CatBoostClassifier,CatBoostRegressor}
+
+function MMI.reformat(::CatBoostModels, X, y)
+    data_pool = prepare_input(X, y)
+    return (data_pool,)
+end
+
+function MMI.reformat(::CatBoostModels, X)
+    x_pool = prepare_input(X)
+    return (x_pool,)
+end
+
+function MMI.selectrows(::CatBoostModels, I, data_pool)
+    py_I = numpy.array(numpy.array(I))
+    return (data_pool.slice(py_I))
```

Comment:
Something wrong here. Why should `data_pool` have `slice` as field? See my other comment. Also, I think your missing a comma:
```suggestion
    return (data_pool.slice(py_I),)
```

Reply:
Should `selectrows` handle data in the format that `MMI.update`/`MMI.fit` uses? I can add methods for handling generic table subsetting too. I have slice here because I assume that the data is in a catboost pool at this point, and you have to slice the pool to subset it. 

https://catboost.ai/en/docs/concepts/python-reference_pool_slice

Reply:
Yes you are right! That's not the issue.  But I have this error, and I think the comma is the problem.

```julia
using MLJ
model = CatBoostClassifier()
mach = machine(model, X, y)
train, test = partition(eachindex(y), 0.7)

julia> fit!(mach, rows=train)
[ Info: Training machine(CatBoostClassifier(iterations = 1000, …), …).
┌ Error: Problem fitting the machine machine(CatBoostClassifier(iterations = 1000, …), …). 
└ @ MLJBase ~/.julia/packages/MLJBase/9kTjj/src/machines.jl:682
[ Info: Running type checks... 
[ Info: Type checks okay. 
ERROR: Python: TypeError: 'Pool' object is not iterable
Python stacktrace: none
Stacktrace:
 [1] pythrow()
   @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/err.jl:94
 [2] errcheck
   @ ~/.julia/packages/PythonCall/3GRYN/src/err.jl:10 [inlined]
 [3] pyiter(x::PythonCall.Py)
   @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/abstract/iter.jl:6
 [4] iterate(x::PythonCall.Py)
   @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/Py.jl:339
 [5] fit_only!(mach::Machine{CatBoostClassifier, true}; rows::Vector{Int64}, verbosity::Int64, force::Bool, composite::Nothing)                                                     
   @ MLJBase ~/.julia/packages/MLJBase/9kTjj/src/machines.jl:680
 [6] #fit!#63
   @ ~/.julia/packages/MLJBase/9kTjj/src/machines.jl:778 [inlined]
 [7] top-level scope
   @ REPL[29]:1
```

Reply:
added the missing comma

Reply:
resolved